### PR TITLE
feat: `rocks upload`

### PR DESCRIPTION
**Pull request by vhyrro**
TODOs:
- [x] DRY for `format!()` calls
- [x] Proper signature creation. In my opinion, signatures should be enabled and enforced by default.

### CODE REVIEW

File: rocks-bin/src/upload.rs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+use eyre::Result;
+use rocks_lib::{config::Config, project::Project, upload::upload_from_project};
+
+pub async fn upload(config: Config) -> Result<()> {
+    let project = Project::current()?.unwrap();
+
+    upload_from_project(
+        &project,
+        env::var("ROCKS_API_KEY")
+            .map_err(|_| eyre!("No API key provided! Please set the $ROCKS_API_KEY variable."))?,
```
Comment (mrcjkb) [inline rocks-bin/src/upload.rs:13]: suggestion(for-another-issue): let's add a config option that allows us to use `libsecret` to fetch API keys,

Comment (mrcjkb) [inline rocks-bin/src/upload.rs:13]: suggestion: We could also prompt for the API key if it's not set.

File: rocks-lib/src/upload/mod.rs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    ToolCheck(#[from] ToolCheckError),
+    UserCheck(#[from] UserCheckError),
+    RockCheck(#[from] RockCheckError),
+    #[error("rock already exists on server: {0}")]
+    RockExists(String),
+}
+
+pub async fn upload_from_project(
+    project: &Project,
+    api_key: String,
```
Comment (mrcjkb) [inline rocks-lib/src/upload/mod.rs:53]: issue: This should be a newtype with a masked `Display` implementation.

File: rocks-lib/src/upload/mod.rs [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        .file_name(format!(
+            "{}-{}.rockspec",
+            rockspec.package, rockspec.version
+        ))
+        .mime_str("application/octet-stream")?;
+    //let signature = reqwest::multipart::Part::text("something").file_name("project.rockspec.sig");
+    let multipart = Form::new().part("rockspec_file", rockspec);
+
+    client
+        .post(format!("{}{}/{}/upload", config.server(), "api/1", api_key))
```
Comment (mrcjkb) [inline rocks-lib/src/upload/mod.rs:106]: question(security): Can we somehow tell the client that the url contains sensitive information?

**Review by mrcjkb — CHANGES_REQUESTED** [review]

File: rocks-lib/src/upload/mod.rs (hunk 3 above)
Comment (vhyrro) [reply rocks-lib/src/upload/mod.rs:106]: Unsure, I'll have to look into that. That's not on us though, I don't know who thought putting the API key in the URL is a good idea :sweat_smile:

File: rocks-bin/src/upload.rs (hunk 1 above)
Comment (vhyrro) [reply rocks-bin/src/upload.rs:13]: I'm personally against this - I'd rather the user drop their API key in a `.env` file or something as opposed to prompting it :)

Comment (mrcjkb) [reply rocks-bin/src/upload.rs:13]: fine by me. I don't feel strongly about it

File: rocks-lib/src/upload/mod.rs (hunk 3 above)
Comment (mrcjkb) [reply rocks-lib/src/upload/mod.rs:106]: If the client doesn't log anything or include the url in error messages, we should be fine.

> I don't know who thought putting the API key in the URL is a good idea

You raise a good point. Firewalls and proxies will likely log the URL unhashed. This is potentially a major security issue that we should raise with luarocks-site.

Comment (vhyrro) [reply rocks-lib/src/upload/mod.rs]: Not sure if it's *major*, but it's definitely larger-than-minor. Any sort of logging immediately leaks the API key :/

Comment (mrcjkb) [reply rocks-lib/src/upload/mod.rs]: which is a risk for supply chain attacks.

File: rocks-lib/src/upload/mod.rs (hunk 2 above)
Comment (vhyrro) [reply rocks-lib/src/upload/mod.rs]: done :)

**Comment by vhyrro** [comment]
question: should I gate `gpgme` behind a feature flag that's enabled by default, or should we bundle GPG in our CI?

**Comment by mrcjkb** [comment]
> question: should I gate `gpgme` behind a feature flag that's enabled by default, or should we bundle GPG in our CI?

We can add `gpg` to the `buildInputs` (or `propagatedBuildInputs`? cc @teto) in the `overlay.nix`.

File: rocks-lib/src/upload/mod.rs (hunk 3 above)
Comment (teto) [reply rocks-lib/src/upload/mod.rs]: it is major omg :o

**Comment by teto** [comment]
sure. Whatever works :)

File: rocks-lib/src/upload/mod.rs (hunk 3 above)
Comment (mrcjkb) [reply rocks-lib/src/upload/mod.rs]: luarocks uses https, so the path will be encrypted. But there is still a risk on the server side.

**Review by mrcjkb — APPROVED** [review]

File: rocks-lib/Cargo.toml [hunk 4]
```diff
@@ -15,9 +15,9 @@ git2 = "0.19.0"
 html-escape = "0.2.13"
 httpdate = "1.0.3"
 itertools = "0.13.0"
-mlua = { version = "0.10.0", features = ["luajit52", "serialize", "send"] }
+mlua = { git = "https://github.com/vhyrro/mlua", features = ["luajit52", "serialize", "send"] }
```
Comment (mrcjkb) [inline rocks-lib/Cargo.toml:18]: question: is this change needed for this feature? Or is it left over from the `rocks add` feature?
If it's needed, we'll have to add

```nix
outputHashes = {
  "mlua-0.10.0" = "<hash>";
};
```
to the cargo lockfile import in the nix build expression.

Comment (vhyrro) [reply rocks-lib/Cargo.toml:18]: It's needed in this PR - I'm wondering whether I could refactor the code to stop this from being a requirement though. I'll look into it in 2 hours.

**Comment by vhyrro** [comment]
Everything's done - all that's left is making the CI happy now.

File: rocks-lib/src/upload/mod.rs [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+            let mut signature = Data::new()?;
+
+            match ctx.sign_detached(rockspec_content.clone(), &mut signature) {
+                Ok(_) => {
+                    let mut signature_str = String::new();
+                    signature.read_to_string(&mut signature_str)?;
+
+                    Some(signature_str)
+                }
+                // TODO: Issue a warning here that signing wasn't successful.
```
Comment (mrcjkb) [inline rocks-lib/src/upload/mod.rs:132]: suggestion: Le'ts not issue a warning. Instead, let's default to returning an `Err`, suggesting to pass in a `--skip-sign` flag.
I think it's better if people have to explicitly skip signing.

File: rocks-lib/src/upload/mod.rs [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+pub async fn upload_from_project(
+    project: &Project,
+    api_key: &ApiKey,
+    config: &Config,
+) -> Result<(), UploadError> {
+    let client = Client::new();
+
+    helpers::ensure_tool_version(&client, config.server()).await?;
+    helpers::ensure_user_exists(&client, api_key, config.server()).await?;
```
Comment (mrcjkb) [inline rocks-lib/src/upload/mod.rs:101]: suggestion: since the API key could be leaked if the URL is not HTTPS, how about we check that it is and return an error if it isn't?

```suggestion

    helpers::ensure_https(config.server())?;
    helpers::ensure_tool_version(&client, config.server()).await?;
    helpers::ensure_user_exists(&client, api_key, config.server()).await?;
```

We could also have `config.server()` perform the check :thinking:

**Review by mrcjkb — APPROVED** [review]

**Comment by mrcjkb** [comment]
> Everything's done - all that's left is making the CI happy now.

I've fixed CI. You'll need to pull the changes in your checkout @vhyrro

File: rocks-lib/src/upload/mod.rs (hunk 5 above)
Comment (vhyrro) [reply rocks-lib/src/upload/mod.rs]: Done. For consistency's sake I made it a `--sign-protocol` flag, allowing the user to select their protocol of choice. Whenever GPG issues an error, the user also receives a hint saying they can pass `--sign-protocol none` to skip the signing step :)

File: rocks-lib/src/upload/mod.rs (hunk 6 above)
Comment (vhyrro) [reply rocks-lib/src/upload/mod.rs:137]: Sounds like a plan. Instead of checking the URL string, I'll see if there's a way to force `reqwest` to make HTTPS calls only.

**Review by mrcjkb — APPROVED** [review]
